Remove non-OSS licenses and source from source-build

ellahathaway opened the pull request:
Closes https://github.com/dotnet/source-build/issues/4736

This PR introduces tooling and functionality to remove non-OSS licenses and associated source in prep-source-build.sh. As is, the tooling removes the following files from the VMR:

```
        deleted:    src/arcade/src/Microsoft.DotNet.Build.Tasks.Installers/build/wix/eula.rtf
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/ANCMIISExpressV2/ancm_iis_expressv2.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/ANCMV2/aspnetcoremodulev2.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/IIS-Setup/iisca/wix/iisca.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/IIS-Setup/iisca/wix3/iisca.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/license/license.rtf
        deleted:    src/aspnetcore/src/Installers/Windows/AspNetCoreModule-Setup/setupstrings.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/HostOptions/Product.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1028/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1029/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1031/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1033/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1036/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1040/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1041/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1042/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1045/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1046/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1049/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/1055/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/2052/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/3082/thm.wxl
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/ANCM.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/Bundle.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/DotNetCore.wxs
        deleted:    src/aspnetcore/src/Installers/Windows/WindowsHostingBundle/SharedFramework.wxs
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.CHS.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.CHT.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.CSY.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.DEU.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.ENU.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.ESN.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.FRA.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.ITA.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.JPN.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.KOR.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.PLK.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.PTB.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.RUS.rtf
        deleted:    src/fsharp/setup/resources/eula/VF_EULA.TRK.rtf
        deleted:    src/roslyn/src/Setup/Roslyn.ThirdPartyNotices/ThirdPartyNotices.rtf
        deleted:    src/roslyn/src/Setup/Roslyn.VsixLicense/EULA.rtf
        deleted:    src/runtime/src/installer/pkg/LICENSE-MSFT.TXT
```

Some things to consider when reviewing the PR:
- The name of the files containing the paths to remove: `disallowed-sb-licenses-paths.txt`. I had a hard time coming up with a name and am open to suggestions on a different one.
- Combining `detect-binaries.sh` and `remove-licenses.sh`? It might be helpful to have a script that shares the common behavior between these two scripts since the two scripts are very similar.
- Adding logging to `FileRemover` tool. If this is done, I'd like to explore sharing the logging infra from `BinaryToolKit` rather than duplicate it in `FileRemover`. This is a potential item to open an issue about and address at a later time.

ellahathaway committed:
Remove non-OSS licenses and source from source-build

ellahathaway commented:
Another item to be done as part of this work is to update the VMR Permissible sources documentation (this PR will go into arcade)

mthalman commented on src/SourceBuild/content/prep-source-build.sh:
```diff
@@ -244,11 +249,21 @@ if [ "$removeBinaries" == true ]; then
+  if [ "$removeBinaries" == true ]; then
+    echo "  Removing non-SB allowed binaries..."
+    "$REPO_ROOT/eng/detect-binaries.sh" \
+    --clean \
```
The `clean` option should not be needed.

ellahathaway replied:
The default of the `detect-binaries.sh` script is to only detect the binaries in the VMR, but not remove them: https://github.com/dotnet/dotnet/blob/main/eng/detect-binaries.sh#L5-L6


As a result, `--clean` is needed here.

mthalman commented on src/SourceBuild/content/eng/tools/FileRemover/Remover.cs:
```diff
@@ -0,0 +1,42 @@
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace FileRemover
+
```
nit: remove blank line

mthalman commented on src/SourceBuild/content/eng/remove-license-files.sh:
I think it would be worthwhile to see what it would look like to have a common implementation with this script and detect-binaries.sh.

ellahathaway replied:
I think that's makes sense, as I'm not a big fan of the level of duplication here either.

I'm thinking I'll follow a similar design pattern to what arcade does with `sdk-tasks`, as seen in [this script](https://github.com/dotnet/arcade/blob/main/eng/common/sdk-task.ps1).

ViktorHofer commented:
As you mentioned, it would be good to merge the scripts and managed implementations. Can happen as part of this change or a follow-up.

MichaelSimons commented on src/SourceBuild/content/eng/disallowed-sb-license-paths.txt:
```diff
@@ -0,0 +1,44 @@
+# Contains the paths to files that contains non-OSS licenses and should be removed from source-build
+
+# This file is used by the FileRemover tool to remove files from the VMR
```
I wonder if "from the VMR" implies too much and if it wouldn't be better to say "locally"?

ellahathaway replied:
Can you explain what you mean by "too much"?

MichaelSimons commented on src/SourceBuild/content/eng/disallowed-sb-license-paths.txt:
Would it be possible to encapsulate the information that a file should be removed within the existing LicenseExclusions.txt file?  The benefit is that it would help make the system more maintainable in that these files would not being specified in multiple places.  I would like to avoid a situation where a new exclusion is added but not stripped.  Maybe there would be a way to update the exclusions file format to help convey the files should be removed (e.g. anything that isn't a false positive should be stripped).

ellahathaway replied:
> Would it be possible to encapsulate the information that a file should be removed within the existing LicenseExclusions.txt file?

_I'm answering this based on the assumption that the excluded files will be removed by prep before license scanning is run._

Yes, this infra is already in place for non-internal branches. 

The license scan tests handle updating the exclusions file (per repo) to remove unused exclusions, and the `CreateBaselineUpdatePR` tool handles combining those exclusions and submitting the PR.

For internal branches, however, the updater tool is turned off. For these cases, I can adjust the tooling to do a "dry-run" (probably based on local sources, but it may be possible use remote sources if I can get OctoKit to pull from Azure devops)? This dry-run would warn in the pipeline if there were differences, but it wouldn't submit a PR. If we want to go this route, I can submit an issue where we can discuss the implementation details in more depth.

MichaelSimons replied:
My two comments are highly related.  Let's consolidate the conversation in https://github.com/dotnet/sdk/pull/45881#discussion_r1913658701.

MichaelSimons commented on src/SourceBuild/content/prep-source-build.sh:
```diff
@@ -46,6 +47,7 @@ buildBootstrap=true
 downloadArtifacts=true
 downloadPrebuilts=true
 removeBinaries=true
+removeLicenses=true
```
Have you thought about how this functionality can be validated?  We should have a test in-place to ensure it doesn't regress.  Would it be possible to run the LicenseScanTests after prep has been run to validate only false-positives are found?

MichaelSimons replied:
My comment was based on looking at how the license scan pipeline was running (e.g. not running prep). If the pipeline starts running prep, then I think that would address my concern.  However, adding prep would raise a UX concern which I reference in https://github.com/dotnet/sdk/pull/45881#discussion_r1913632961.  If a new non-OSS license is found a PR would get opened to add an exclusion in the LicenseScanTests baseline file.  I think it would be helpful to describe (e.g. update the PR description) what the UX is in regards to the LicenseScanTests, new/removed OSS files, and auto-generated PRs to update baseline/exclusion files.

ellahathaway replied:
>  I think it would be helpful to describe (e.g. update the PR description) what the UX is in regards to the LicenseScanTests, new/removed OSS files, and auto-generated PRs to update baseline/exclusion files.

This makes sense. Like you suggested, I think updating the PR description may help to address this UX issue. Maybe adding something like:

```
Follow these guidelines for how to update this PR appropriately:

- Allowed open-source licenses → list of allowed licenses in `LicenseScanTests.cs`
- Files types that shouldn't be scanned (e.g., image files) → list of excluded file types in `LicenseScanTests.cs`
- False positives → `LicenseExclusions.txt`
- Source-build removable non-OSS licenses and source (e.g., test data and Windows installers) → `disallowed-sb-license-paths.txt`
- Cloak other non-OSS licenses and sources
```

ellahathaway committed:
Update src/SourceBuild/content/eng/remove-license-files.sh

Co-authored-by: Matt Thalman <mthalman@microsoft.com>

ellahathaway commented on src/SourceBuild/content/eng/remove-license-files.sh:
```diff
@@ -0,0 +1,129 @@
+      print_help
+      exit 0
+      ;;
+    --clean)
+      mode="clean"
+      ;;
```
Ah yeah. Copy paste error. Will fix.

ellahathaway commented on src/SourceBuild/content/eng/disallowed-sb-license-paths.txt:
Yes. I pulled these from the `LicenseExclusions.txt` file. Based on my understanding, the list of files I've included in `disallowed-sb-license-paths.txt` are non-OSS licenses included in the VMR for installers. It's possible that we only want to strip the windows installers, in which case some of the paths I've included can be removed from this list of files.

ellahathaway commented:
Closing this in favor of addressing the comments/feedback in other PRs. See https://github.com/dotnet/source-build/issues/4736#issuecomment-2591005939 for more details.
